packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,489 @@
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount): Promise<void> {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+    const releaseLock = await this.#mutex.acquire();
+
+    try {
+      // Get assets list
+      if (account.metadata.snap) {
+        const assets = await this.#getAssetsList(
+          account.id,
+          account.metadata.snap.id,
+        );
+        await this.#refreshAssetsMetadata(assets);
+        this.update((state) => {
+          state.allNonEvmTokens[account.id] = assets;
+        });
+      }
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    // Check if accountId is in allNonEvmTokens and if it is, remove it
+    if (this.state.allNonEvmTokens[accountId]) {
+      this.update((state) => {
+        delete state.allNonEvmTokens[accountId];
+      });
+    }
+  }
+
+  /**
+   * Refreshes the assets snaps and metadata for the given list of assets
+   *
+   * @param assets - The assets to refresh
+   */
+  async #refreshAssetsMetadata(assets: CaipAssetType[]) {
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const chainIdResponse = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainIdResponse.chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainIdResponse = parseCaipAssetType(asset);
+      if (!assetsByScope[chainIdResponse.chainId]) {
+        assetsByScope[chainIdResponse.chainId] = [];
+      }
+      assetsByScope[chainIdResponse.chainId].push(asset);
+    }
+    let newMetadata: Record<CaipAssetType, FungibleAssetMetadata> = {};
+    for (const chainId in assetsByScope) {
+      if (hasProperty(assetsByScope, chainId)) {
```

Comment:
Feels unnecessary to me? We are already iterating over `assetsByScope` so this property will always be defined?

Reply:
I did that because eslint was not happy to make the for-in loop without the guard
`The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.`

Reply:
Oh actually, I might now why 🙈

It should be `of` and not `in` (unlike python 😄)

This seems to work fine:

```ts
for (const scope of Object.keys(assetsByScope)) {
  ...
```

Also, we could use `.values()` here maybe, since we are not using `scope` (`chainId` in your code) at all, WDYT?

Reply:
Right, safe to use `for (const chainId of Object.keys(assetsByScope)) {` and remove the check  `if (hasProperty(assetsByScope, chainId))` here [1887ff5](https://github.com/MetaMask/core/pull/5138/commits/1887ff5b7c951e15b7bcbd535f3fd03fb7e43cc0)